src/sentry/tasks/integrations/slack/link_slack_user_identities.py

```diff
@@ -66,3 +87,38 @@ def link_slack_user_identities(
                     "type": idp.type,
                 },
             )
+
+
+# has different typing for slack_data_by_user, need to grab slack_id from a dataclass
+def update_identities_with_dataclass(
+    slack_data_by_user: Mapping[User, SlackUserData], idp: IdentityProvider
+):
+    date_verified = timezone.now()
+    identities_by_user = get_identities_by_user(idp, slack_data_by_user.keys())
+
+    for user, data in slack_data_by_user.items():
+        slack_id = data.slack_id
+        # Identity already exists, the emails match, AND the external ID has changed
+        if user in identities_by_user.keys():
+            if slack_id != identities_by_user[user].external_id:
+                # replace the Identity's external_id with the new one we just got from Slack
+                identities_by_user[user].update(external_id=data.slack_id)
+            continue
+        # the user doesn't already have an Identity and one of their Sentry emails matches their Slack email
+        matched_identity, created = Identity.objects.get_or_create(
+            idp=idp,
+            external_id=slack_id,
+            defaults={"user": user, "status": IdentityStatus.VALID, "date_verified": date_verified},
+        )
+        # the Identity matching that idp/external_id combo is linked to a different user
+        if not created:
+            logger.info(
+                "post_install.identity_linked_different_user",
+                extra={
+                    "idp_id": idp.id,
+                    "external_id": slack_id,
+                    "object_id": matched_identity.id,
+                    "user_id": user.id,
+                    "type": idp.type,
+                },
+            )
```

Comment:
This is a lot of duplicated code from `update_identities`. It would be better to adapt the existing function to work with both cases.

There are a few ways we could go about that. My first choice would be to rewrite the existing function to act on `SlackUserData` objects, and convert the `Mapping[str, str]` values from old function calls into the new dataclass. If that won't work for some reason, we could also refactor it into a very small abstract class.

Let me know if you'd like to pair on that.

Reply:
it's copy-pasted, using the new dataclass for the old logic sounds good to me